posthog/models/session_replay_event/sql.py

```diff
@@ -0,0 +1,118 @@
+from django.conf import settings
+
+from posthog.clickhouse.kafka_engine import kafka_engine
+from posthog.clickhouse.table_engines import Distributed, ReplicationScheme, AggregatingMergeTree
+from posthog.kafka_client.topics import KAFKA_CLICKHOUSE_SESSION_REPLAY_EVENTS
+
+SESSION_REPLAY_EVENTS_DATA_TABLE = lambda: "sharded_session_replay_events"
+
+"""Kafka needs slightly different column setup. It receives individual events, not aggregates."""
+KAFKA_SESSION_REPLAY_EVENTS_TABLE_BASE_SQL = """
+CREATE TABLE IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
+(
+    session_id VARCHAR,
+    team_id Int64,
+    distinct_id VARCHAR,
+    first_timestamp DateTime64(6, 'UTC'),
+    last_timestamp DateTime64(6, 'UTC'),
+    first_url Nullable(VARCHAR),
+    click_count Int64,
+    keypress_count Int64,
+    mouse_activity_count Int64
+) ENGINE = {engine}
+"""
+
+SESSION_REPLAY_EVENTS_TABLE_BASE_SQL = """
+CREATE TABLE IF NOT EXISTS {table_name} ON CLUSTER '{cluster}'
+(
+    session_id VARCHAR,
+    team_id Int64,
+    distinct_id VARCHAR,
+    first_timestamp AggregateFunction(min, DateTime64(6, 'UTC')),
+    last_timestamp AggregateFunction(max, DateTime64(6, 'UTC')),
+    first_url Nullable(VARCHAR),
+    click_count SimpleAggregateFunction(sum, Int64),
+    keypress_count SimpleAggregateFunction(sum, Int64),
+    mouse_activity_count SimpleAggregateFunction(sum, Int64)
+) ENGINE = {engine}
+"""
+
+
+SESSION_REPLAY_EVENTS_DATA_TABLE_ENGINE = lambda: AggregatingMergeTree(
+    "session_replay_events", replication_scheme=ReplicationScheme.SHARDED
+)
+
+SESSION_REPLAY_EVENTS_TABLE_SQL = lambda: (
+    SESSION_REPLAY_EVENTS_TABLE_BASE_SQL
+    + """ORDER BY (team_id, session_id)
+SETTINGS index_granularity=512
```

Comment:
Can we partition the table by... hmmm... month maybe? That at least puts some bounds on the size of the table and we can maybe perform further optimizations on old partitions that we know won't be written to: https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/custom-partitioning-key

Reply:
I had partition by timestamp at first but because it's an aggregate column I couldn't use it for that... or at least couldn't figure out how to

Reply:
Maybe we partition by hash of session id?

Reply:
hmm, interesting. I'm sure there's a way to make this work 🤔 

hash of session id doesn't really give us the temporal element however? Ideally we want to know partition x isn't been written to, let's completely optimize it, or maybe we drop it, or maybe we perform a schema migration partition wise.

Reply:
🥳